internal/jobobject/jobobject.go

```diff
@@ -546,3 +541,45 @@ func (job *JobObject) PromoteToSilo() error {
 func (job *JobObject) isSilo() bool {
 	return atomic.LoadUint32(&job.isAppSilo) == 1
 }
+
+// PrivateWorkingSet returns the private working set size for the job. This is calculated by adding up the
+// private working set for every process running in the job.
+func (job *JobObject) PrivateWorkingSet() (uint64, error) {
+	pids, err := job.Pids()
+	if err != nil {
+		return 0, err
+	}
+
+	openAndQuery := func(pid uint32) (uint64, error) {
+		h, err := windows.OpenProcess(windows.PROCESS_QUERY_LIMITED_INFORMATION, false, pid)
+		if err != nil {
+			return 0, fmt.Errorf("failed to open process with pid %d: %w", pid, err)
+		}
+		defer func() {
+			_ = windows.Close(h)
+		}()
+
+		var vmCounters winapi.VM_COUNTERS_EX2
+		status := winapi.NtQueryInformationProcess(
+			h,
+			winapi.ProcessVmCounters,
+			uintptr(unsafe.Pointer(&vmCounters)),
+			uint32(unsafe.Sizeof(vmCounters)),
+			nil,
+		)
+		if !winapi.NTSuccess(status) {
+			return 0, fmt.Errorf("failed to query information for process with pid %d: %w", pid, winapi.RtlNtStatusToDosError(status))
+		}
+		return uint64(vmCounters.PrivateWorkingSetSize), nil
+	}
+
+	var jobWorkingSetSize uint64
+	for _, pid := range pids {
```

Comment:
There is some risk that a PID could be reused between when we call `Pids` and when we open that process. Can we check process start time or something to ensure we're looking at the same processes?

Reply:
Good point, looks like HCS handles this case as well. That'll make this a bit slower but shouldn't be too terrible I'd hope

Reply:
Thinking more about this, I think the reason HCS handles this is might be more for some of the other queries. By virtue of using NtQuerySystemInformation they get all the data upfront and aren't cycling through one by one, so once they have their snapshot of info they don't need to (well, there's no instance that they do) re-query afterwards. This is troublesome for the ProcessList query in HCS for example as some of the data returned from this query *is* the creation time and image name, so they'd need to re-query again if this didn't match up. I guess for us, we're taking the role of NtQuerySystemInfo which requires us to open the process first ourselves to grab this data, so we don't have anything to compare against before opening. We could take a timestamp from right after Pids() returns and see if a creation time is after this I guess? 

Looks like they also don't exit out on failure to open the process which we should probably do however (probably to handle if a process exited in between querying and open).

Reply:
Going to add a timestamp before we call pids and check that when we open the proc its creation time isn't after this

Reply:
Added, ptal

Reply:
Wouldnt it be just as safe to do:

pids = GetPids()
for each pid in pids
    handles <- open handle // skip of fails to open

for each handle in handles 
    jobWSS += query working set from handle
    closehandle(handle)

return

Reply:
IE: open all pid handles so they cant be closed without you. Then query them after the fact. Then close all handles.

Reply:
Yea that'd probably cut down on the time that a pid could get re-used in comparison to the:
```
pids = GetPids()
for each pid in pids
    p = open(pid)
    stats = query(p)
```
that's in this change now. The timestamp approach definitely isn't flawless, but I suppose a process could exit and get its pid re-used as we're cycling through and opening the handles in your approach as well. I think the main idea is we'd rather miss a process that was started in the job slightly after the timestamp we take than potentially report a process that's not running in the job and just exited and had its pid re-used

Reply:
One new idea based on what Justin said: We could get the job's PIDs, and then enumerate, open each pid, and check if it's still in the job using `IsProcessInJob`. As Justin said, the PID won't be re-used while we have a handle open to the process.

Reply:
Oh, for whatever reason I remembered `IsProcessInJob` just telling you if it was in a job in general and not a specific one.. Yea that works fine.

Reply:
@kevpar @jterry75 Done (after some fun stack corruption issues with IsProcessInJob), ptal.

Reply:
I like this approach.

Reply:
Yep LGTM